### PR TITLE
enable AWS dispatcher tests

### LINKED ISSUE #492: BUG dispatcher failing on lack of instance
Another error related to AWS.

```
Traceback (most recent call last):
  File "/opt/miniconda3/bin/ramp-launch", line 8, in <module>
    sys.exit(start())
  File "/opt/miniconda3/lib/python3.7/site-packages/ramp_engine/cli.py", line 122, in start
    main()
  File "/opt/miniconda3/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/opt/miniconda3/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/opt/miniconda3/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/miniconda3/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/miniconda3/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/opt/miniconda3/lib/python3.7/site-packages/ramp_engine/cli.py", line 90, in dispatcher
    disp.launch()
  File "/opt/miniconda3/lib/python3.7/site-packages/ramp_engine/dispatcher.py", line 309, in launch
    self.launch_workers(session)
  File "/opt/miniconda3/lib/python3.7/site-packages/ramp_engine/dispatcher.py", line 171, in launch_workers
    worker.teardown()  # kill the worker
  File "/opt/miniconda3/lib/python3.7/site-packages/ramp_engine/aws/worker.py", line 190, in teardown
    self.config, self.instance.id
AttributeError: 'AWSWorker' object has no attribute 'instance'
```

### PR DESCRIPTION
**Pull request by maikia**
As for now the AWS worker is poorly tested. 

This PR:
1. adds the necessary tools to test the AWS from the dispatcher level (ie creates the aws event in the test database session)
2. it adds a sample test for not launched instance due to authorization issue
3. it closes #492 by checking if the instance is assigned to the worker before killing the worker
4. it also cleans up the tests of aws worker to use the same type of template (and places it in the same locations) as it is done for the conda worker

sorry for such a long PR. It was hard to cut it once it started to grow :-/

### CODE REVIEW

**Comment by maikia** [comment]
@rth I would be curious about your opinion on the config file.
Here I am placing the `submissions_dir`, `predictions_dir` and `logs_dir` in both worker and ramp for local and remote separately. This might need to be clarified in the docs if you agree with that approach.

File: ramp-database/ramp_database/tests/data/ramp_config_aws_iris.yml [hunk 1]
```diff
@@ -3,9 +3,12 @@ ramp:
...
     event_title: "iris aws"
     event_is_public: true
-    data_dir: ramp-board/ramp_deployment/ramp-data/iris
-    kit_dir: ramp-board/ramp-engine/ramp_engine/tests/kits
+    data_dir: /tmp/databoard_test/ramp-data/iris
+    kit_dir: /tmp/databoard_test/ramp-kits/iris
     sandbox_dir: starting_kit
+    submissions_dir: /tmp/databoard_test/submissions
+    predictions_dir: /tmp/databoard_test/preds
+    logs_dir: /tmp/databoard_test/log
```
Comment (agramfort) [inline ramp-database/ramp_database/tests/data/ramp_config_aws_iris.yml:11]: maybe a bit dangerous that the system as /tmp folder?

**Review by agramfort — COMMENTED** [review]
thanks a ton @maikia !

did you spot any issue doing this?

File: ramp-database/ramp_database/tests/data/ramp_config_aws_iris.yml (hunk 1 above)
Comment (maikia) [reply ramp-database/ramp_database/tests/data/ramp_config_aws_iris.yml:11]: hm. I used the exactly the same directories for `submissions`, `preds` and `log` as it is already set for `ramp_config_iris.yml` (ie iris running on conda). `boston_housing` is also using `/tmp` folder. If you have alternative suggestion we should change it for all three

File: ramp-engine/ramp_engine/tests/test_dispatcher.py [hunk 2]
```diff
@@ -216,3 +231,23 @@ def test_dispatcher_worker_retry(session_toy):
 
     submissions = get_submissions(session_toy, 'iris_test', 'new')
     assert submission_name in [sub[1] for sub in submissions]
+
+
+def test_dispatcher_aws_not_launching(session_toy_aws, caplog):
+    # given the test config file the instance should not be able to launch
+    # on authentication error
```
Comment (lucyleeow) [inline ramp-engine/ramp_engine/tests/test_dispatcher.py:238]:
```suggestion
    # due to authentication error
```
?

**Review by lucyleeow — APPROVED** [review]
Very useful to have the tests!

**Comment by maikia** [comment]
@tomMoral would you mind having a look? thx

File: ramp-database/ramp_database/model/tests/test_submission.py [hunk 3]
```diff
@@ -299,7 +311,10 @@ def test_submission_score_on_cv_fold_model_property(session_scope_module):
 def test_submission_on_cv_fold_model_property(session_scope_module):
     cv_fold = \
         (session_scope_module.query(SubmissionOnCVFold)
-                             .filter(SubmissionOnCVFold.submission_id == 5)
+                             .filter(
+                                 SubmissionOnCVFold.submission_id ==
+                                 ID_SUBMISSION
+                                 )
```
Comment (tomMoral) [inline ramp-database/ramp_database/model/tests/test_submission.py:317]:
```suggestion
                             )
```

File: ramp-database/ramp_database/model/tests/test_submission.py [hunk 4]
```diff
@@ -327,7 +342,10 @@ def test_submission_on_cv_fold_model_is_public_leaderboard(
         session_scope_module, state_set, expected_state):
     cv_fold = \
         (session_scope_module.query(SubmissionOnCVFold)
-                             .filter(SubmissionOnCVFold.submission_id == 5)
+                             .filter(
+                                 SubmissionOnCVFold.submission_id ==
+                                 ID_SUBMISSION
+                                 )
```
Comment (tomMoral) [inline ramp-database/ramp_database/model/tests/test_submission.py:348]:
```suggestion
                             )
```

File: ramp-database/ramp_database/model/tests/test_submission.py [hunk 5]
```diff
@@ -352,7 +370,10 @@ def test_submission_on_cv_fold_model_is_trained(session_scope_module,
                                                 state_set, expected_state):
     cv_fold = \
         (session_scope_module.query(SubmissionOnCVFold)
-                             .filter(SubmissionOnCVFold.submission_id == 5)
+                             .filter(
+                                 SubmissionOnCVFold.submission_id ==
+                                 ID_SUBMISSION
+                                 )
```
Comment (tomMoral) [inline ramp-database/ramp_database/model/tests/test_submission.py:376]:
```suggestion
                             )
```

File: ramp-database/ramp_database/tools/tests/test_submission.py [hunk 6]
```diff
@@ -65,6 +65,7 @@
 from ramp_database.tools.submission import submit_starting_kits
 
 HERE = os.path.dirname(__file__)
+ID_SUBMISSION = 7
```
Comment (tomMoral) [inline ramp-database/ramp_database/tools/tests/test_submission.py:68]: Is it supposed to be the same as the one in `ramp-database/ramp_database/model/tests/test_submission.py`?
If yes, maybe define it uniquely in `ramp-database/ramp_database/testing.py`

File: ramp-engine/ramp_engine/tests/test_dispatcher.py [hunk 7]
```diff
@@ -216,3 +231,23 @@ def test_dispatcher_worker_retry(session_toy):
...
+                            event_config=event_config,
+                            worker=AWSWorker, n_workers=10,
+                            hunger_policy='exit')
+    dispatcher.fetch_from_db(session_toy_aws)
+    dispatcher.launch_workers(session_toy_aws)
+    assert 'AuthFailure' in caplog.text
+    # training should not have started
+    assert 'training' not in caplog.text
+    num_running_workers = dispatcher._processing_worker_queue.qsize()
+    assert num_running_workers == 0
```
Comment (tomMoral) [inline ramp-engine/ramp_engine/tests/test_dispatcher.py:253]: I suppose there are some subimissions in this db?
Is there a way to make sure that the submission is still in a proper `new` state even though the worker did not start?

**Review by tomMoral — APPROVED** [review]
A couple of cosmit and one question to extend the tests but overall it looks pretty good!
Thx @maikia

File: ramp-database/ramp_database/tools/tests/test_submission.py (hunk 6 above)
Comment (maikia) [reply ramp-database/ramp_database/tools/tests/test_submission.py:68]: could be... but I think it's easier to debug if you have the variable in the same testing file.. 
if it's very important to you I will change but I am slightly leaning against it.

Comment (tomMoral) [reply ramp-database/ramp_database/tools/tests/test_submission.py:68]: It is not that important. It is just that when you have constant, I find it better to not have multiple definitions. But if it is not critical (as it is in the tests), do what you think is the best :)

**Comment by maikia** [comment]
@tomMoral could you please merge if you are ok with the changes?
Thx

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thanks @maikia and @tomMoral !

**Pull request merged**